compiler/src/dotty/tools/dotc/interactive/Interactive.scala

```diff
@@ -269,38 +269,69 @@ object Interactive {
    *
    *  @param includeReferences  If true, include references and not just definitions
    */
   def namedTrees(trees: List[SourceTree], includeReferences: Boolean, treePredicate: NameTree => Boolean)
-    (implicit ctx: Context): List[SourceTree] = safely {
-    val buf = new mutable.ListBuffer[SourceTree]
+    (implicit ctx: Context): List[SourceNamedTree] = safely {
+    val buf = new mutable.ListBuffer[SourceNamedTree]
 
-    trees foreach { case SourceTree(topTree, source) =>
+    def traverser(source: SourceFile) = {
       new untpd.TreeTraverser {
+        private def handleImport(imported: List[Symbol],
+                                 uexpr: untpd.Tree,
+                                 id: untpd.Ident,
+                                 rename: Option[untpd.Ident]): Unit = {
+          val expr = uexpr.asInstanceOf[tpd.Tree]
+          imported.foreach { sym =>
+            val tree = tpd.Select(expr, sym.name).withPos(id.pos)
+            val renameTree = rename.map { r =>
+              val name = if (sym.name.isTypeName) r.name.toTypeName else r.name
+              RenameTree(name, tpd.Select(expr, sym.name)).withPos(r.pos)
+            }
+
+            traverse(tree)
+            renameTree.foreach(traverse)
+          }
+
+          traverse(expr)
+        }
         override def traverse(tree: untpd.Tree)(implicit ctx: Context) = {
-          tree match {
+        tree match {
+            case imp @ Import(uexpr, (id: untpd.Ident) :: Nil) =>
+              val imported = importedSymbols(imp.asInstanceOf[tpd.Import])
+              handleImport(imported, uexpr, id, None)
+            case imp @ Import(uexpr, Thicket((id: untpd.Ident) :: (rename: untpd.Ident) :: Nil) :: Nil) =>
+              val imported = importedSymbols(imp.asInstanceOf[tpd.Import])
+              handleImport(imported, uexpr, id, Some(rename))
             case utree: untpd.NameTree if tree.hasType =>
               val tree = utree.asInstanceOf[tpd.NameTree]
               if (tree.symbol.exists
-                   && !tree.symbol.is(Synthetic)
-                   && tree.pos.exists
-                   && !tree.pos.isZeroExtent
-                   && (includeReferences || isDefinition(tree))
-                   && treePredicate(tree))
-                buf += SourceTree(tree, source)
+                && !tree.symbol.is(Synthetic)
+                && tree.pos.exists
+                && !tree.pos.isZeroExtent
+                && (includeReferences || isDefinition(tree))
+                && treePredicate(tree))
+                  buf += SourceNamedTree(tree, source)
               traverseChildren(tree)
             case tree: untpd.Inlined =>
               traverse(tree.call)
             case _ =>
               traverseChildren(tree)
           }
         }
-      }.traverse(topTree)
+      }
+    }
+
+    trees foreach {
+      case SourceNamedTree(topTree, source) =>
+        traverser(source).traverse(topTree)
```

Comment:
Looks like the two cases do the same thing, why not `trees.foreach(tree => traverser(source).traverse(tree.tree))` ?